planning/route_handler/src/route_handler.cpp

```diff
@@ -712,110 +710,134 @@ lanelet::ConstLanelets RouteHandler::getLaneletSequence(
+      road_lanelets_at_pose.push_back(lanelet_at_pose);
   }
-  return false;
+  return road_lanelets_at_pose;
 }
 
-bool RouteHandler::getLeftShoulderLanelet(
-  const lanelet::ConstLanelet & lanelet, lanelet::ConstLanelet * left_lanelet) const
+std::optional<lanelet::ConstLanelet> RouteHandler::getFollowingShoulderLanelet(
+  const lanelet::ConstLanelet & lanelet) const
 {
-  for (const auto & shoulder_lanelet : shoulder_lanelets_) {
-    if (lanelet::geometry::leftOf(shoulder_lanelet, lanelet)) {
-      *left_lanelet = shoulder_lanelet;
-      return true;
-    }
+  bool found = false;
+  const auto & search_point = lanelet.centerline().back().basicPoint2d();
+  const auto next_lanelet = lanelet_map_ptr_->laneletLayer.nearestUntil(
+    search_point, [&](const auto & bbox, const auto & ll) {
+      if (isShoulderLanelet(ll) && lanelet::geometry::follows(lanelet, ll)) found = true;
+      // stop search once next shoulder lanelet is found, or the bbox does not touch the search
+      // point
+      return found || lanelet::geometry::distance2d(bbox, search_point) > 1e-3;
+    });
+  if (found && next_lanelet.has_value()) return *next_lanelet;
+  return std::nullopt;
+}
+
+std::optional<lanelet::ConstLanelet> RouteHandler::getLeftShoulderLanelet(
+  const lanelet::ConstLanelet & lanelet) const
+{
+  for (const auto & other_lanelet :
+       lanelet_map_ptr_->laneletLayer.findUsages(lanelet.leftBound())) {
+    if (other_lanelet.rightBound() == lanelet.leftBound() && isShoulderLanelet(other_lanelet))
+      return other_lanelet;
   }
-  return false;
+  return std::nullopt;
 }
 
-bool RouteHandler::getRightShoulderLanelet(
-  const lanelet::ConstLanelet & lanelet, lanelet::ConstLanelet * right_lanelet) const
+std::optional<lanelet::ConstLanelet> RouteHandler::getRightShoulderLanelet(
+  const lanelet::ConstLanelet & lanelet) const
 {
-  for (const auto & shoulder_lanelet : shoulder_lanelets_) {
-    if (lanelet::geometry::rightOf(shoulder_lanelet, lanelet)) {
-      *right_lanelet = shoulder_lanelet;
-      return true;
-    }
+  for (const auto & other_lanelet :
+       lanelet_map_ptr_->laneletLayer.findUsages(lanelet.rightBound())) {
+    if (other_lanelet.leftBound() == lanelet.rightBound() && isShoulderLanelet(other_lanelet))
+      return other_lanelet;
   }
-  return false;
+  return std::nullopt;
+}
+
+std::optional<lanelet::ConstLanelet> RouteHandler::getShoulderLaneletAtPose(const Pose & pose) const
+{
+  const lanelet::BasicPoint2d p{pose.position.x, pose.position.y};
+  const auto lanelets_at_pose = lanelet_map_ptr_->laneletLayer.search(lanelet::BoundingBox2d(p));
+  for (const auto & lanelet_at_pose : lanelets_at_pose) {
+    // confirm that the pose is inside the lanelet since "search" does an approximation with boxes
+    const auto is_pose_inside_lanelet = lanelet::geometry::inside(lanelet_at_pose, p);
+    if (is_pose_inside_lanelet && isShoulderLanelet(lanelet_at_pose)) return lanelet_at_pose;
```

Comment:
The function description includes 
```
/// @warning if the pose is inside multiple shoulder lanelets, the selection is random, 
```
which caught my attention. I have a question out of curiosity, not as a change request.

If the following conditions are met:
- shoulder lanes overlap
- pose is located in an area where shoulder lanes overlap
Is it possible to retrieve multiple shoulder lanes that satisfy these conditions by using loop?

Copy

Reply:
Thank you for checking the PR.
> Is it possible to retrieve multiple shoulder lanes that satisfy these conditions by using loop?

No it is not possible, "random" was probably the wrong word.
To avoid confusion, I updated the function to return all shoulder lanelets: 1945b6cda13dd811c07280eddbbf17a998eeca85

Reply:
Thank you so much for explanation!
I prefer updated implementation! Thanks!